src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs

```diff
@@ -58,12 +55,27 @@ protected void UnregisterText(SourceTextContainer textContainer)
 
             var registration = GetWorkspaceRegistration(textContainer);
 
+            // guard us from being called with wrong text container.
             if (registration.Workspace == this)
             {
-                registration.SetWorkspaceAndRaiseEvents(null);
```

Comment:
register text is async event, and unregister as synchronous events which can cause ordering issue (unregister fire first and then register event get fired) so moved all events to async